lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,210 @@
+defmodule Kino.SmartCell.ChartBuilder do
+  use Kino.JS, assets_path: "lib/assets/chart_builder"
+  use Kino.JS.Live
+  use Kino.SmartCell, name: "Chart builder"
+
+  @as_int ["width", "height"]
+  @as_atom ["data_variable", "chart_type", "x_field_type", "y_field_type", "color_field_type"]
+
+  @impl true
+  def init(attrs, ctx) do
+    fields = %{
+      "chart_type" => attrs["chart_type"] || "bar",
+      "width" => attrs["width"],
+      "height" => attrs["height"],
+      "x_field" => attrs["x_field"],
+      "y_field" => attrs["y_field"],
+      "x_field_type" => attrs["x_field_type"],
+      "y_field_type" => attrs["y_field_type"],
+      "color_field" => attrs["color_field"],
+      "color_field_type" => attrs["color_field_type"],
+      "data_variable" => attrs["data_variable"]
+    }
+
+    ctx =
+      assign(ctx,
+        fields: fields,
+        options: %{},
+        vl_alias: nil,
+        missing_dep: missing_dep(),
+        fresh: attrs == %{}
+      )
+
+    {:ok, ctx, reevaluate_on_change: true}
+  end
+
+  @impl true
+  def scan_binding(pid, binding, env) do
+    data_options =
+      for {key, val} <- binding, is_atom(key), is_map(val), into: %{}, do: {key, Map.keys(val)}
+
+    vl_alias = vl_alias(env)
+    send(pid, {:scan_binding_result, data_options, vl_alias})
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      fields: ctx.assigns.fields,
+      missing_dep: ctx.assigns.missing_dep,
+      options: ctx.assigns.options,
+      fresh: ctx.assigns.fresh
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, data_options, vl_alias}, ctx) do
+    ctx = assign(ctx, options: data_options, vl_alias: vl_alias)
+    broadcast_event(ctx, "set_available_data", %{"options" => data_options})
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => "data_variable", "value" => ""}, ctx) do
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"field" => "data_variable", "value" => value}, ctx) do
+    updated_fields = update_data_options(ctx, value)
+    ctx = update(ctx, :fields, &Map.merge(&1, updated_fields))
+    broadcast_event(ctx, "update", %{"fields" => updated_fields})
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"field" => field, "value" => value}, ctx) do
+    parsed_value = parse_value(field, value)
+    ctx = update(ctx, :fields, &Map.put(&1, field, parsed_value))
+    broadcast_event(ctx, "update", %{"fields" => %{field => parsed_value}})
+
+    {:noreply, ctx}
+  end
+
+  defp update_data_options(ctx, value) do
+    df = String.to_atom(value)
+    axis = ctx.assigns.options[df] |> List.first()
+    axis = if axis, do: Atom.to_string(axis)
```

Comment:
Would it make more sense to default x and y to different columns?

```suggestion
    {x_field, y_field} =
      case ctx.assigns.options[df] do
        [] -> {nil, nil}
        [key] -> {Atom.to_string(key), Atom.to_string(key)}
        [key1, key2 | _] -> {Atom.to_string(key1), Atom.to_string(key2)}
      end
```

Also, we could convert the column names to strings upfront in `scan_binding`, since we don't need them as atoms.